src/backend/distributed/executor/multi_router_executor.c

```diff
@@ -69,12 +87,63 @@ RouterExecutorStart(QueryDesc *queryDesc, int eflags, Task *task)
 	/* ensure that the task is not NULL */
 	Assert(task != NULL);
 
-	/* disallow transactions and triggers during distributed modify commands */
+	/* disallow triggers during distributed modify commands */
 	if (commandType != CMD_SELECT)
 	{
-		bool topLevel = true;
-		PreventTransactionChain(topLevel, "distributed commands");
 		eflags |= EXEC_FLAG_SKIP_TRIGGERS;
+
+		if (IsTransactionBlock() && xactParticipantHash == NULL)
```

Comment:
> Hm. Not that I have a better idea, but this implies that we'll not enter this block when called from within a function :(

I agree with this (and didn't have a better idea, other than some ugly hacks that would work but be slow).

> Hm. What about checking `QueryDesc->dest`, and treating everything but `DestRemote`, `DestRemoteExecute` as commands needing a transaction? I think that should do what we want.

_This_ I do not understand. Can you clarify what you're proposing? It sounds clean and like it could handle the other "multi-statements-inside-transactions" cases (functions and multi-command strings), but I don't know what `QueryDesc->dest` has to do with those cases. Quick explanation?